### ISSUE TITLE
Snapped onto the motorway

### ISSUE THREAD

**Issue by Guo-astro** [valhalla/valhalla#1820]
When wanna going to a place near the motorway,  the target point  is snapped on the motorway, and we need to stop driving on the motorway which is very dangerous and prohibited by the law.
Example:
curl http://localhost:8002/route --data '{"locations":[{"lat":35.169715,"lon":136.931980},{"lat":35.194333,"lon": 136.911871}],"costing":"auto","costing_options":{"auto":{"use_tolls":0}}
,"directions_options":{"units":"km"}}'

ANy idea to avoid this?

**Comment by kevinkreiser** [comment]
you could add a new feature to the `auto` `costing_options`. have a look in `sif` at `autocost.cc`. there you can add a new option like `maximum_candidate_road_class`. it should be a string which you can map to an enum of values. once you have added it to the protobuf (directions_options costing_options) you can then look at `edge_filter` in `autocost.cc`. In there you can have a bit of code that "filters" edges that have a road class higher than the ones you want to allow. Please note that this filter only applies to the snapping (loki) phase of route generation.

just a little bit of code like:

```c++
  virtual const EdgeFilter GetEdgeFilter() const {
    // Throw back a lambda that checks the access for this type of costing
    return [this](const baldr::DirectedEdge* edge) {
      // note that RoadClass has important roads with low numbers
      if (edge->is_shortcut() || !(edge->forwardaccess() & kAutoAccess) || maximum_candidate_road_class > edge->classification()) {
        return 0.0f;
      } else {
        // TODO - use classification/use to alter the factor
        return 1.0f;
      }
    };
  }
```

note that the classifications are as follows:

```c++
enum class RoadClass : uint8_t {
  kMotorway = 0,
  kTrunk = 1,
  kPrimary = 2,
  kSecondary = 3,
  kTertiary = 4,
  kUnclassified = 5,
  kResidential = 6,
  kServiceOther = 7
};
```

**Comment by Guo-astro** [comment]
@kevinkreiser  Thanks very much for the clear response! I had done that!  Should I PR after issue #1815 merged?

**Comment by kevinkreiser** [comment]
yes please! ill take another look at #1815 in the next couple of days. sorry for the delay!

**Comment by danpaz** [comment]
@Guo-astro did you get a chance to start this issue? I am exploring a similar feature and wondering if you made any progress here.

**Comment by danpaz** [comment]
A potential solution to this problem is moving in https://github.com/valhalla/valhalla/pull/2289. Should allow you to include in `search_filters:{"max_road_class":"trunk"}` on the destination location if you want to avoid snapping to motorways.
